File: survival-c/game.c
#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#include "game.h"
#include "error.h"

bool controls[4] = {false, false, false, false};

frames_t currentFrame = FRONT_STANDING;

Frames front = { .frame_type = FRONT, .frames = { FRONT_WALK_RIGHT_HAND, FRONT_STANDING, FRONT_WALK_LEFT_HAND }, .next_frame = FRONT_WALK_LEFT_HAND };
Frames left = { .frame_type = LEFT, .frames = { LEFT_WALK_RIGHT_HAND, LEFT_WALK_STANDING, LEFT_WALK_LEFT_HAND }, .next_frame = LEFT_WALK_LEFT_HAND };
Frames right = { .frame_type = RIGHT, .frames = { RIGHT_WALK_RIGHT_HAND, RIGHT_WALK_STANDING, RIGHT_WALK_LEFT_HAND }, .next_frame = RIGHT_WALK_LEFT_HAND };
Frames back = { .frame_type = BACK, .frames = { BACK_WALK_RIGHT_HAND, BACK_WALK_STANDING, BACK_WALK_LEFT_HAND }, .next_frame = BACK_WALK_LEFT_HAND };

// TODO: implement
// when enemy hits the player, player looses life
// if life <= 0, game over
// when enemy hits, we can make the player faster for a few secs
// for now it only returns true or false and we kill the player if it hits on main.c
bool enemyHit(Player *p, Enemy *e) {
  if (p->position.x < e->position.x + ENEMY_WIDTH &&
      p->position.x + SPRITE_WIDTH > e->position.x &&
      p->position.y < e->position.y + ENEMY_HEIGHT &&
      p->position.y + SPRITE_HEIGHT > e->position.y) {
    return true;
  }
  return false; 
}

bool hitAnyEnemy(Player *p, Enemy **enemies) {
  for (int i = 0; i < NUM_ENEMIES; i++) {
    if (!enemies[i]->dead && enemyHit(p, enemies[i])) {
      return true;
    }
  }
  return false;
}

void updateEnemyPosition(Player *p, Enemy *e) {
  Position direction = {
    p->position.x - e->position.x,
    p->position.y - e->position.y
  };

  float length = sqrt(direction.x * direction.x + direction.y * direction.y);
  direction.x /= length;
  direction.y /= length;

  e->position.x += direction.x * e->speed;
  e->position.y += direction.y * e->speed;

  if (fabs(direction.x) > fabs(direction.y)) {
    if (direction.x > 0) {
      e->currentFrame = 11;
    } else {
      e->currentFrame = 6;
    }
  } else {
    if (direction.y > 0) {
      e->currentFrame = 3;
    } else {
      if (direction.x > 0) {
        e->currentFrame = 11;
      } {
        e->currentFrame = 6;
      }
    }
  }
}

void updateEnemiesPosition(Player *p, Enemy **enemies) {
  for (int i = 0; i < NUM_ENEMIES; i++) {
    if (!enemies[i]->dead) {
      updateEnemyPosition(p, enemies[i]);
    }
  }
}


Enemy *createEnemyAtRandomPos(int life, int damage, int speed) {
  Enemy *enemy = malloc(sizeof(Enemy));
  if (enemy == NULL) {
    error("Error allocating memory for enemy\n");
  }

  enemy->position.x = rand() % (WINDOW_WIDTH - ENEMY_WIDTH);
  enemy->position.y = rand() % (WINDOW_HEIGHT - ENEMY_HEIGHT);
  enemy->life = life;
  enemy->damage = damage;
  enemy->speed = speed;
  enemy->dead = false;
  enemy->currentFrame = FRONT_STANDING;

  return enemy;
}

Enemy** createEnemies(int numEnemies, int life, int damage, int speed) {
  Enemy **enemies = malloc(sizeof(Enemy*) * numEnemies);
  if (enemies == NULL) {
    error("Error allocating memory for enemies array\n");
  }
  for (int i = 0; i < numEnemies; i++) {
    enemies[i] = createEnemyAtRandomPos(life, damage, speed);
  }
  return enemies;
}

int updateFrontFrame(frames_t current) {
  int curr = (int)current;
  if (curr >= FRONT_WALK_LEFT_HAND) {
    return FRONT_WALK_RIGHT_HAND;
  }
  return curr + 1;
}

int updateLeftFrame(frames_t current) {
  int curr = (int)current;
  if (curr >= LEFT_WALK_LEFT_HAND) {
    return LEFT_WALK_RIGHT_HAND;
  }
  return curr + 1;
}

int updateRightFrame(frames_t current) {
  int curr = (int)current;
  if (curr >= RIGHT_WALK_LEFT_HAND) {
    return RIGHT_WALK_RIGHT_HAND;
  }
  return curr + 1;
}

int updateBackFrame(frames_t current) {
  int curr = (int)current;
  if (curr >= BACK_WALK_LEFT_HAND) {
    return BACK_WALK_RIGHT_HAND;
  }
  return curr + 1;
}

void updateNextFrame(frames_t f) {
  currentFrame = f;
}

frames_t getCurrentFrame() {
  return currentFrame;
}

frames_t getNextFrame(controls_t control) {
  frames_t f;
  switch (control) {
    case W:
      f = back.next_frame;
      back.next_frame = updateBackFrame(f);
      return f;
    case S:
      f = front.next_frame;
      front.next_frame = updateFrontFrame(f);
      return f;
    case A:
      f = left.next_frame;
      left.next_frame = updateLeftFrame(f);
      return f;
    case D:
      f = right.next_frame;
      right.next_frame = updateRightFrame(f);
      return f;
  }
  return FRONT_STANDING;
}

bool isControlPressed(controls_t control) {
  return controls[control];
}

void pressControl(controls_t control) {
  controls[control] = true;
}

void releaseControl(controls_t control) {
  controls[control] = false;
}

void updatePlayerPosition(Player *p) {
    if (isControlPressed(W)) {
        p->position.y -= SPEED;
    }
    if (isControlPressed(S)) {
        p->position.y += SPEED;
    }
    if (isControlPressed(A)) {
        p->position.x -= SPEED;
    }
    if (isControlPressed(D)) {
        p->position.x += SPEED;
    }
}

void clampPlayerPosition(Player *p) {
    if (p->position.x < 0) p->position.x = 0;
    if (p->position.y < 0) p->position.y = 0;
    if (p->position.x > WINDOW_WIDTH - SPRITE_WIDTH) p->position.x = WINDOW_WIDTH - SPRITE_WIDTH;
    if (p->position.y > WINDOW_HEIGHT - SPRITE_HEIGHT) p->position.y = WINDOW_HEIGHT - SPRITE_HEIGHT;
}

bool hit(Player *p, Object *o) {
  if (p->position.x < o->position.x + o->width &&
      p->position.x + SPRITE_WIDTH > o->position.x &&
      p->position.y < o->position.y + o->height &&
      p->position.y + SPRITE_HEIGHT > o->position.y) {
    return true;
  }
  return false; 
}

void clampObjectCollision(Player *p, Object *o) {
    if (hit(p, o)) {
        // Calculate the overlap on each axis
        float overlapX = (p->position.x < o->position.x + o->width / 2) ?
            (p->position.x + SPRITE_WIDTH) - o->position.x :
            (o->position.x + o->width) - p->position.x;
        
        float overlapY = (p->position.y < o->position.y + o->height / 2) ?
            (p->position.y + SPRITE_HEIGHT) - o->position.y :
            (o->position.y + o->height) - p->position.y;

        // Determine which axis has the smaller overlap and adjust only that axis
        if (overlapX < overlapY) {
            if (p->position.x < o->position.x + o->width / 2) {
                p->position.x = o->position.x - SPRITE_WIDTH;
            } else {
                p->position.x = o->position.x + o->width;
            }
        } else {
            if (p->position.y < o->position.y + o->height / 2) {
                p->position.y = o->position.y - SPRITE_HEIGHT;
            } else {
                p->position.y = o->position.y + o->height;
            }
        }
    }
}

File: survival/src/Game.hs
module Game where

import Foreign.C.Types (CInt)
import Data.Array
import Data.Ord (comparing)
import Data.List (maximumBy)

windowWidth :: CInt
windowWidth = 950

windowHeight :: CInt
windowHeight = 800

spriteWidth :: Float
spriteWidth = 26

spriteHeight :: Float
spriteHeight = 36

frameRows :: Float
frameRows = 5

frameCols :: Float
frameCols = 3

animationFrames :: Float
animationFrames = frameRows * frameCols

enemyWidth :: Float
enemyWidth = 36

enemyHeight :: Float
enemyHeight = 36

enemyCols :: Float
enemyCols = 3

enemyRows :: Float 
enemyRows = 4

enemyAnimationFrames :: Float
enemyAnimationFrames = enemyCols * enemyRows

numEnemies :: Float
numEnemies = 7

speed :: Float
speed = 2

fps :: Int
fps = 60

frameDelay :: Int
frameDelay = 1000 `div` fps

data LoopState = Quit | Running | Won | Lost deriving (Show, Enum, Bounded, Eq)

data Controls = W | A | S | D deriving (Show, Enum, Bounded, Eq, Ord)

data FramesType = 
    FRONT_WALK_RIGHT_HAND
  | FRONT_STANDING
  | FRONT_WALK_LEFT_HAND
  | LEFT_WALK_RIGHT_HAND
  | LEFT_WALK_STANDING
  | LEFT_WALK_LEFT_HAND
  | RIGHT_WALK_RIGHT_HAND
  | RIGHT_WALK_STANDING
  | RIGHT_WALK_LEFT_HAND
  | BACK_WALK_RIGHT_HAND
  | BACK_WALK_STANDING
  | BACK_WALK_LEFT_HAND
  deriving (Show, Enum, Bounded, Eq)

data FrameDirection = FRONT | LEFT | RIGHT | BACK deriving (Show, Enum, Bounded, Eq, Ord)

data Frames = Frames
  { frameDirection :: FrameDirection
  , frames ::[FramesType]
  , nextFrame :: FramesType
  } deriving (Show)

data Position = Position
  { x :: Float
  , y :: Float
  } deriving (Show, Eq)

data Enemy = Enemy
  { ePosition :: Position
  , eLife :: Int
  , damage :: Int
  , eSpeed :: Float
  , eDead :: Bool
  , currentFrame :: FramesType 
  } deriving (Show, Eq)

data Player = Player
  { pPosition :: Position
  , life :: Int
  , pSpeed :: Float
  , pDead :: Bool
  , pCurrentFrame :: FramesType
  } deriving (Show, Eq)

controls :: Array Int Bool
controls = array (0, 3) [(i, False) | i <- [0..3]]

enemyHit :: Player -> Enemy -> Bool
enemyHit player enemy =
          let  px = x (pPosition player)
               py = y (pPosition player)
               ex = x (ePosition enemy )
               ey = y (ePosition enemy ) 
          in px < ex + enemyWidth  &&
             px + spriteWidth > ex &&
             py < ey + enemyHeight &&
             py + spriteHeight > ey


hitAnyEnemy :: Player -> [Enemy] -> Bool
hitAnyEnemy player enemies = any (enemyHit player) enemies

-- return the enemy object with the position updated. The calculu
updateEnemyPosition :: Player -> Enemy -> Enemy
updateEnemyPosition p e =
  let direction = Position (x (pPosition p) - x (ePosition e)) (y (pPosition p) - y (ePosition e))
      lengthDist = sqrt (x direction * x direction + y direction * y direction)
      normalizedDirection = Position ((x direction) / lengthDist) ((y direction) / lengthDist)
      newPosition = Position
        ((x (ePosition e)) + (x normalizedDirection) * (eSpeed e))
        ((y (ePosition e)) + (y normalizedDirection) * (eSpeed e))
      newFrame = determineFrame normalizedDirection
   in e { ePosition = newPosition , currentFrame = newFrame }


-- return the next correct frame to the animation
determineFrame :: Position -> FramesType
determineFrame dir
  | abs (x dir) > abs (y dir) = if (x dir) > 0 then BACK_WALK_LEFT_HAND else RIGHT_WALK_RIGHT_HAND
  | (y dir) > 0 = LEFT_WALK_RIGHT_HAND
  | (x dir) > 0 = BACK_WALK_LEFT_HAND
  | otherwise = RIGHT_WALK_RIGHT_HAND


updateEnemiesPosition :: Player -> [Enemy] -> [Enemy]
updateEnemiesPosition p = map (updateEnemyPosition p) . filter (not . eDead)

pseudoRandom :: Int -> Int
pseudoRandom seed = (1103515245 * seed + 12345) `mod` 2147483648

createEnemyAtRandomPos :: Int -> Int -> Int -> Float -> Enemy
createEnemyAtRandomPos seed life damage speed =
  let
    windowWidth = 950
    windowHeight = 800
    randX = pseudoRandom seed
    randY = pseudoRandom (seed + 1)

    enemyWidth = 36
    enemyHeight = 36

    posX = fromIntegral (randX `mod` (windowWidth - enemyWidth))
    posY = fromIntegral (randY `mod` (windowHeight - enemyHeight))
  in
    Enemy
      { ePosition = Position posX posY
      , eLife = life
      , damage = damage
      , eSpeed = speed
      , eDead = False
      , currentFrame = FRONT_STANDING
      }

-- this functions create N enemies in different positions, with the same life, damage and speed
createEnemies :: Int -> Int -> Int -> Float -> [Enemy]
createEnemies numEnemies life damage speed = 
  let seeds = [1..numEnemies]
  in map (\seed -> createEnemyAtRandomPos seed life damage speed) seeds

updateFrontFrame :: FramesType -> FramesType
updateFrontFrame FRONT_WALK_LEFT_HAND = FRONT_WALK_RIGHT_HAND
updateFrontFrame FRONT_WALK_RIGHT_HAND = FRONT_STANDING
updateFrontFrame FRONT_STANDING = FRONT_WALK_LEFT_HAND
updateFrontFrame other = other

updateLeftFrame :: FramesType -> FramesType
updateLeftFrame LEFT_WALK_LEFT_HAND = LEFT_WALK_RIGHT_HAND
updateLeftFrame LEFT_WALK_RIGHT_HAND = LEFT_WALK_STANDING
updateLeftFrame LEFT_WALK_STANDING = LEFT_WALK_LEFT_HAND
updateLeftFrame other = other

updateRightFrame :: FramesType -> FramesType
updateRightFrame RIGHT_WALK_LEFT_HAND = RIGHT_WALK_RIGHT_HAND
updateRightFrame RIGHT_WALK_RIGHT_HAND = RIGHT_WALK_STANDING
updateRightFrame RIGHT_WALK_STANDING = RIGHT_WALK_LEFT_HAND
updateRightFrame other = other

updateBackFrame :: FramesType -> FramesType
updateBackFrame BACK_WALK_LEFT_HAND = BACK_WALK_RIGHT_HAND
updateBackFrame BACK_WALK_RIGHT_HAND = BACK_WALK_STANDING
updateBackFrame BACK_WALK_STANDING = BACK_WALK_LEFT_HAND
updateBackFrame other = other
































